classifier/main.py

```diff
@@ -126,6 +127,9 @@ def _classify_with_snn(alert_dict: dict) -> dict:
     # classify
     _, pred_probs = classify_lcs(snn_df, model_path, device)
 
+    kafka_timestamp = datetime.fromtimestamp(int(attrs["kafka.timestamp"])/1000)
+    formatted_kafka_timestamp = kafka_timestamp.strftime("%Y-%m-%d %H:%M:%S.%f") + " UTC"
+
```

Comment:
Is all of this format conversion really necessary? This is a timestamp in the BigQuery schema, so it seems like we ought to be able to just send the int.

Reply:
I wanted to avoid the formatting as well but for consistency, I think it is necessary. The `brokerIngestTimestamp` and `classifierTimestamp` data entries specify the timezone (UTC). Should I just send the int?

Reply:
Timestamps are typically floats and they are independent of timezone. Read about:
- [Unix Timestamp](https://www.unixtimestamp.com/) (scroll down to "What is the unix time stamp?")
- [BigQuery Timestamp type](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp_type)

> seems like we ought to be able to just send the int.

Correction, I meant float not int. I think we should be able to send `float(attrs["kafka.timestamp"])`.

We are dealing with the other two timestamps as datetimes, and datetimes require timezones. Datetimes are coercible to timestamps so I'm guessing BigQuery is doing the conversion under the hood.  We have to do a conversion on the `brokerIngestTimestamp` because it comes in as a string (this will be cleaned up in #20).

Reply:
(For learning purposes, look more closely at your conversion, particularly:

```
int(attrs["kafka.timestamp"])/1000
```

I think you are incorrectly altering the timestamp.)

Reply:
Sorry, need to clean up my ints/floats again. Timestamps in general may be ints or floats. I believe that the kafka.timestamp is a float (represented as a string) but you should check. BigQuery Timestamps have microsecond precision, so I think that means we should just send the kafka.timestamp as a float.

Reply:
> 
> We are dealing with the other two timestamps as datetimes, and datetimes require timezones. Datetimes are coercible to timestamps so I'm guessing BigQuery is doing the conversion under the hood. We have to do a conversion on the `brokerIngestTimestamp` both because it comes in as a string and because the outgoing Avro schema has requirements for it (this will be cleaned up in #20).

Ah, I see. Thank you for clarifying. I had originally planned on implementing a similar approach as to what's done in the `_create_elasticc_msg` function, where `elasticcPublishTimestamp = int(attrs["kafka.timestamp"])` and the `elasticcPublishTimestamp` value is passed in the `msg` (dictionary). But as I was messing around with the `datetime` module on my terminal (ipython session), I noticed that the "UTC" was missing... and well it got me to the point I'm at now. I'll double check the type (int/float) of kafka.timestamp, make changes, test them, and report back

Reply:
I don't remember where the line `elasticcPublishTimestamp = int(attrs["kafka.timestamp"])` came from or if I ever checked whether we really do need `int` here rather than `float`. If kafka.timestamp really is a float, then we should try to send it in the message as a float as well (otherwise, we're losing precision). But it might be that the Avro schema restricts this to an int, idk.

Reply:
I kept wondering if I was remembering wrong, so just did a quick check and indeed...

kafka.timestamp is an integer (represented as a string).  So sending it as an `int` in the message makes sense. Let's also try to send it to BigQuery as an `int`.

Reply:
Thanks for tracking some of this down. I'm sorry for all the confusion I caused by thinking that kafka.timestamp was a float. I've looked at the info you provided and done more of my own testing and I think I have it sorted out now 🤞...

The value we should send to BigQuery is:

```
int(attrs["kafka.timestamp"])/1000
```

That is a float in units of seconds, with millisecond precision. A timestamp is a number -- it's not a datetime or a string. That is a general statement about timestamps and is why my initial reaction was that all the conversions seemed unnecessary. (To understand what BigQuery is doing and what you're seeing in the console, read [BigQuery Timestamp type](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp_type) and pay attention to how it is _stored_ vs. _displayed_.)

The tricky bit here is figuring out the units and precision. The kafka.timestamp is an integer in units of milliseconds (as you learned). We must send this to BigQuery in units of seconds (as I learned by trial and error... the docs are frustratingly mute about this) and BigQuery stores it with microsecond precision (as stated in the docs).

Your original solution was:

> ```
> kafka_timestamp = datetime.fromtimestamp(int(attrs["kafka.timestamp"])/1000)
> formatted_kafka_timestamp = kafka_timestamp.strftime("%Y-%m-%d %H:%M:%S.%f") + " UTC"
> ```

The first line is fine except that it returns a datetime in the local timezone, so appending the string "UTC" in the second line is incorrect. We could convert it to UTC and pass `formatted_kafka_timestamp` (a string) to BigQuery... BigQuery would just convert the string back into a timestamp to store it. Better to avoid all the unnecessary conversions since they make the code harder to read and easy to introduce errors.

Reply:
sounds good to me